### PR TITLE
Implemented refresh endpoint for google provider

### LINKED ISSUE #993: Auth backend: Handle refresh token flow

### PR DESCRIPTION
**Pull request by soapraj**
- handles token refresh for google in the auth backend and frontend apis

#### :heavy_check_mark: Checklist

- [ ] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [ ] Prettier run on changed files
- [ ] Tests added for new functionality
- [ ] Regression tests added for bug fixes

### CODE REVIEW

File: plugins/auth-backend/src/providers/google/provider.ts [hunk 1]
```diff
@@ -54,15 +56,61 @@ export class GoogleAuthProvider
...
+      const options: CookieOptions = {
+        maxAge: THOUSAND_DAYS_MS,
+        secure: false,
+        sameSite: 'none',
+        domain: 'localhost',
+        path: '/auth/google',
+        httpOnly: true,
+      };
+
+      res.cookie('grtoken', refreshToken, options);
```
Comment (Rugvip) [inline plugins/auth-backend/src/providers/google/provider.ts:76]: Should use the provider id or smth to build a unique name for these cookie

File: plugins/auth-backend/src/providers/google/provider.ts [hunk 2]
```diff
@@ -54,15 +56,61 @@ export class GoogleAuthProvider
...
+      return postMessageResponse(res, {
         type: 'auth-result',
         payload: user,
       });
     })(req, res, next);
   }
 
   async logout(_req: express.Request, res: express.Response) {
-    res.send('logout!');
+    return res.send('logout!');
```
Comment (Rugvip) [inline plugins/auth-backend/src/providers/google/provider.ts:85]: Clear the cookie here

File: plugins/auth-backend/src/providers/google/provider.test.ts [hunk 3]
```diff
@@ -110,34 +112,107 @@ describe('GoogleAuthProvider', () => {
...
+      const googleAuthProvider = new GoogleAuthProvider(
+        googleAuthProviderConfig,
+      );
+
+      googleAuthProvider.frameHandler(mockRequest, mockResponse, mockNext);
+      expect(spyPassport).toBeCalledTimes(1);
+      expect(spyPostMessage).toBeCalledTimes(1);
+      expect(spyPostMessage).toBeCalledWith(mockResponse, {
+        type: 'auth-result',
+        error: new Error('Google auth failed Error: TokenError'),
```
Comment (freben) [inline plugins/auth-backend/src/providers/google/provider.test.ts:214]: missing a comma here :)

File: plugins/auth-backend/src/providers/google/provider.ts [hunk 4]
```diff
@@ -25,6 +26,7 @@ import {
 import { postMessageResponse } from './../utils';
 import { InputError } from '@backstage/backend-common';
 
+export const THOUSAND_DAYS_MS = 1000 * 24 * 60 * 60 * 1000;
```
Comment (freben) [inline plugins/auth-backend/src/providers/google/provider.ts:29]: OK BUT WHAT ABOUT LEAP SECONDS

File: plugins/auth-backend/src/providers/google/provider.ts [hunk 5]
```diff
@@ -53,16 +55,87 @@ export class GoogleAuthProvider
     res: express.Response,
     next: express.NextFunction,
   ) {
-    return passport.authenticate('google', (_, user) => {
-      postMessageResponse(res, {
+    return passport.authenticate('google', (err, user) => {
+      if (err) {
+        return postMessageResponse(res, {
+          type: 'auth-result',
+          error: new Error(`Google auth failed ${err.toString()}`),
```
Comment (freben) [inline plugins/auth-backend/src/providers/google/provider.ts:62]: `Google auth failed, ${err}`

File: plugins/auth-backend/src/providers/google/provider.ts [hunk 6]
```diff
@@ -53,16 +55,87 @@ export class GoogleAuthProvider
...
+          type: 'auth-result',
+          error: new Error('Missing refresh token'),
+        });
+      }
+
+      delete user.refreshToken;
+
+      const options: CookieOptions = {
+        maxAge: THOUSAND_DAYS_MS,
+        secure: false,
```
Comment (freben) [inline plugins/auth-backend/src/providers/google/provider.ts:79]: for now? eventually make configurable, and/or plain depend on NODE_ENV right?

File: plugins/auth-backend/src/providers/google/provider.ts [hunk 7]
```diff
@@ -53,16 +55,87 @@ export class GoogleAuthProvider
...
         payload: user,
       });
     })(req, res, next);
   }
 
   async logout(_req: express.Request, res: express.Response) {
-    res.send('logout!');
+    const options: CookieOptions = {
+      maxAge: 0,
+      secure: false,
```
Comment (freben) [inline plugins/auth-backend/src/providers/google/provider.ts:101]: same here

File: plugins/auth-backend/src/providers/google/provider.ts [hunk 8]
```diff
@@ -53,16 +55,87 @@ export class GoogleAuthProvider
...
+        });
+      }
+
+      delete user.refreshToken;
+
+      const options: CookieOptions = {
+        maxAge: THOUSAND_DAYS_MS,
+        secure: false,
+        sameSite: 'none',
+        domain: 'localhost',
```
Comment (freben) [inline plugins/auth-backend/src/providers/google/provider.ts:81]: for now?

File: plugins/auth-backend/src/providers/google/provider.ts (hunk 5 above)
Comment (Rugvip) [inline plugins/auth-backend/src/providers/google/provider.ts:62]: This won't be properly serialized in the response, but I think we can defer the fix for that to the refactor

**Review by freben — APPROVED** [review]
Cool! No real complaints

File: plugins/auth-backend/src/providers/google/provider.ts (hunk 4 above)
Comment (Rugvip) [reply plugins/auth-backend/src/providers/google/provider.ts:29]: ITS DAYS, NO YEARS

File: plugins/auth-backend/src/providers/google/provider.ts (hunk 6 above)
Comment (soapraj) [reply plugins/auth-backend/src/providers/google/provider.ts:79]: yes. this and domain will change based on NODE_ENV but the whole config and dev/prod setup would be a separate PR later.

Comment (Rugvip) [reply plugins/auth-backend/src/providers/google/provider.ts:79]: We need to support both environments simultaneously in the backend, it should be switched based on request parameters

**Pull request merged**